refactor

mixmix opened the pull request:
**changes**
- i reworked some of the encoding/decoding to lean heavily on the `bfe.json` definitions.
- `throw` when something fails to match a decode/encode path
    - feels better to know something is ill defined than to just jam it through and likely produce junk?
- moved the `sigil` property should be on the type
    - every TF with a sigil has the same sigil
    - ... hmmm but maybe not in future? will see


 
I've done my best to group processing of encoding/ decoding into clear cases, and also be careful not to mint new buffers by comparing slices (I _think_ that uses pointers and doesn't use more memory?)

**outcomes**:
- turned up some inconsistencies in `bfe.json` which I've since patched.
- revealed edge cases
    - encoding a Buffer leaves it untouched, but then trying to decode the same buffer... you are likely gonna get :fire: :poop:
    - some things have dangerously similar suffixes - `.sig.ed25519` vs `.ed25519`
- realised
    - every TF with a `sigil` has a `suffix` (tested)
    - if something has `suffix`, but no `sigil` the `suffix` is unique (tested)



interested to hear what you both think

mixmix commented on bfe.json:
```diff
@@ -2,52 +2,55 @@
     "type": "signature",
     "formats": [
-      { "code": 0, "format": "ed25519", "signature_length": 64 }
+      { "code": 0, "format": "ed25519", "data_length": 64, "signature_length": 64, "suffix": ".sig.ed25519" }
```
I stopped short or removing unique `X_length` keys, as this would be a breaking change.
would like to see them go though

mixmix replied:
I mean I think the key should always be data_length

mixmix commented on index.js:
```diff
@@ -2,271 +2,169 @@
-    return Buffer.concat([tf, d])
+  uri(input) {
+    let [type, format, ...data] = input.slice(6).split('/')
+    // *FIRE* WARNING ... are we URIEncoding our URIs?!
```
oh yeah ... this. I guess we're discussing URIs on monday...

mixmix replied:
URI decode/ encode removed from this PR for now

staltz commented on index.js:
```diff
@@ -2,271 +2,169 @@
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+  uri(input) {
```
This PR is named "refactor" so it made me think it's just about shuffling things around, but it also turns out it's adding new features. I think this should have been made more clear, because it needs tests, and it needs to adhere to the ssb-uri-spec, see more notes below

mixmix replied:
URI decode/ encode removed from this PR for now

staltz commented on index.js:
```diff
@@ -2,271 +2,169 @@
-
-    return Buffer.concat([tf, d])
+  uri(input) {
+    let [type, format, ...data] = input.slice(6).split('/')
```
The data part is known to not have `/`, so we shouldn't do an array spread here. In fact this code is missing treatment of `_` and `+`. All the `_` and `+` have to be replaced with (respectively) `/` and `-` **before** doing a `Buffer.from(....., 'base64')`

mixmix replied:
Didn't know this!

staltz replied:
2nd paragraph under Canonical: https://github.com/ssb-ngi-pointer/ssb-uri-spec#canonical-ssb-uris

mixmix replied:
URI decode/ encode removed from this PR for now

staltz commented on index.js:
```diff
@@ -2,271 +2,169 @@
+  classic(input, type, format) {
+    let data = input
+    if (type.sigil) data = data.slice(1)
+    if (format.suffix) data = data.replace(format.suffix, '')
```
`data.slice(0, -format.suffix.length)` would be safer here because we know for sure that the removal is happening at the end of the string. This code is probably safe, but not future-proof (who knows if there would be a suffix `foo` in the future which can also randomly occur in the middle of the base64 string).

mixmix replied:
oh didn't know you could do this, that's awesome

staltz commented on index.js:
```diff
@@ -2,271 +2,169 @@
-    return Buffer.concat([STRING_TF, d])
-  },
+  if (typeof input === 'string') {
+    if (input.startsWith('ssb://')) return encoder.uri(input)
```
Not all SSB URIs have `//` some of them may be just `ssb:feed/sha256/`, see https://github.com/staltz/ssb-uri2/blob/5bb4c01ba8ba244c850c3e502ac635651d0dfebc/test/fixtures.js#L13-L15

mixmix replied:
Good to know. Honestly there are multiple variants of this spec and I didn't know which to follow.
So yeah I'll pull the uri stuff out into a different pr + make it better tested

mixmix replied:
removed

staltz commented on package.json:
```diff
@@ -19,6 +19,7 @@
   },
   "scripts": {
     "test": "tape test/*.js | tap-bail | tap-spec",
+    "test:all": "tape test/*.js",
```
What's the point of this one?

mixmix replied:
Oh, that's me not wanting to tap-bail. I was also using tap-diff on output to spot differences in long strings. Can tidy this up

mixmix commented:
@staltz thanks for the review. I've tidied this up:
- removed uri stuff, making this more refactor focused
- rebased (and squashed) to one commit

NOTE: this still has some breaking changes - throwing when it doesn't now how to handle a case. I'll go through and comment on those to highlight them

mixmix commented:
Wrote tests around current throwing behaviour : https://github.com/ssb-ngi-pointer/ssb-bfe/pull/18/files
Rebased this on top of it.

Will leave some comments about modifications have made

mixmix commented on util.js:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  // if it's reached here it's not a known .box* suffix
+  if (isEncrypted(input)) throw new Error('Unknown encrypted format')
```
Added this catch - this assumes that `<base64>.box<n>` was meant to be encrypted type, but if it's not a known type it throws

arj03 commented on index.js:
```diff
@@ -1,300 +1,147 @@
   decode,
-  bfeTypes: TYPES,
-  bfeNamedTypes: NAMED_TYPES,
+  toString: decode,
```
can we leave decode here? I kind of like it :) Fine with adding an alias for toString

mixmix replied:
yeah decode is still there :)

arj03 commented on test/bfe.js:
```diff
@@ -0,0 +1,36 @@
+  t.equal(
+    sigiledTypes.length,
+    new Set(sigiledTypes.map((type) => type.sigil)).size, // unique sigil
+    'each sigil is unique to a type'
```
👍

arj03 commented on index.js:
```diff
@@ -1,300 +1,147 @@
+    const d = input ? BOOL.TRUE : BOOL.FALSE
+    return Buffer.concat([BOOL.TF, d])
   },
+  classic(input, type, format) {
```
I keep looking at this name and thinking it has something to do with classic format. At least as long as we still have this kind of classic encoding of other feed formats. Maybe `classicSigilSuffix`? No so good with names.

staltz replied:
Couple of suggestions:

- `sigilSuffix`
- `ref`
- `sigilRef`

mixmix replied:
DONE

arj03 commented on index.js:
```diff
@@ -1,300 +1,147 @@
-
-    const b64part = d.toString('base64')
-    return '@' + b64part + feedExtension
+  classic(input, type, format) {
```
same here

arj03 commented on util.js:
```diff
@@ -0,0 +1,92 @@
+      undefined
+}
+
+function sortByNames(types) {
```
ohh, this doesn't sort by name :) It converts it into a dict ;)

mixmix replied:
good spotting!

mixmix replied:
renamed `definitionsToDict`

arj03 commented on util.js:
```diff
@@ -0,0 +1,92 @@
+    return { type, format }
+  }
+
+  // for (const type of types) {
```
this should be removed

arj03 commented:
/end of review

This is a great refactor and logical now that we have information encoded in bfe.json so we don't have it 2 places. I mainly have some naming suggestions otherwise this looks good.

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
-
-    return Buffer.concat([tf, d])
+  string(input) {
+    return Buffer.concat([Buffer.from([6, 0]), Buffer.from(input, 'utf8')])
```
This should be using `STRING.TF`, not the hard-coded bytes

mixmix replied:
Great spotting, this is much better

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
-
-  return NAMED_TYPES
+const STRING = {
+  TF: Buffer.from([6, 0]),
```
There's quite a lot of back and forth going on with refactors.

- [It used to be `Buffer.from([6, 0])`](https://github.com/ssb-ngi-pointer/ssb-bfe/blob/609c7e095c9f357b123552fdc8e8c4dca0dcc49b/index.js#L24)
- [We started using bfe.json to fetch the `code`s](https://github.com/ssb-ngi-pointer/ssb-bfe/blob/a097a14b1f5f914641d096fcb2ca9d943159ea77/index.js#L55)
- Now this PR is reverting to how it was in the beginning

I believe you could use `definitions` or `TYPES` here, right?

Further, I don't see a need to change `STRING_TF` to `STRING.TF`, is this just stylistic? Why not preserve the previous style?

mixmix replied:
yeah it's totally styalistic. I think spacially some and like grouping things, and dots make much more sense to me that heaps of SNAKE_CASE_DESCRIPTIONS. I don't know what I find that slower to parse. I think for STRING it doesn't make that much sense, but I really like it for the BOOL case, and was trying to be consistent.

mixmix replied:
I reverted it because I didn't understand the advantage of using magic strings over magic numbers. The problem with magic strings is if it was types wrong, you can _nothing_ out of the object.

I don't care that much about this change at the end of the day. Happy to revert if you prefer

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
+    const { type, format } = findClassicTypeFormat(input, TYPES)
+    if (type) {
+      if (format) return encoder.classic(input, type, format)
+      else throw new Error(`Unknown ${type.type} format`)
```
```suggestion
      else throw new Error(`Unknown ${format} format`)
```

mixmix replied:
Don't agree.
This is meant to output e.g. "Unknown feed format".  The type.type = "feed" here.

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
-    return input
   }
+
+  throw new Error('cannot encoding, type is not defined')
```
In all the errors, we have been trying to provide the faulty data, so the error is useful in a debugging context. I suggest changing this one to tell what the faulty data is.

mixmix replied:
agree, will change

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
+    ].join('')
   },
+  bool(input) {
+    if (input.size > 3) throw new Error('boolean BFE must be 3 bytes')
```
```suggestion
    if (input.size > 3) throw new Error('boolean BFE must be 3 bytes, was ' + input.size)
```

mixmix replied:
agree

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
-  },
+  if (Buffer.isBuffer(input)) {
+    if (input.length < 2)
+      throw new Error('Buffer is missing first two type & format bytes')
```
There was nothing wrong with the previous error we had in place for this, which was:

```suggestion
      throw new Error(
        'Buffer is missing first two type&format fields: ' + input
      )
```

mixmix replied:
Agree, will revert

staltz commented on index.js:
```diff
@@ -1,300 +1,147 @@
+    if (input.equals(NIL.TFD)) return null
+    if (input.slice(0, 2).equals(BOOL.TF)) return decoder.bool(input)
+    if (input.slice(0, 2).equals(STRING.TF))
+      return input.slice(2).toString('utf8')
```
I liked how it was before: there was a decoder object with `decoder.string` which decoding details for each case. Now this is bringing them back, all concerns mixed together.

mixmix replied:
agree. Have moved this into `decoder.string`

staltz commented:
@mixmix Reviewing this PR is taking a lot of time and energy, and there's too much going on. The [feedback I gave here applies directly](https://github.com/ssb-js/ssb-ref/pull/45#issuecomment-895867250).

Could I ask for another try, on a fresh new PR could you do **only** the following changes that you described initially?

> lean on the `bfe.json` definitions
>
> moved the sigil property should be on the type

If you keep it limited to only those kinds of decisions, it should be a rather small PR to write and small to review too.

mixmix commented:
Here's a PR which moves the sigils stuff : https://github.com/ssb-ngi-pointer/ssb-bfe/pull/21/files (merged)

mixmix commented:
Here's part 2, which is the refactor to use bfe.json : https://github.com/ssb-ngi-pointer/ssb-bfe/pull/22
